### ISSUE TITLE
FEATURE support NER pipeline? and How

### ISSUE THREAD

**Issue by shuhliu7** [explosion/spacy-transformers#23]
Hi, 
I am wondering whether the current codebased support ner pipeline from spacy, thanks.

**Comment by JulianGerhard21** [comment]
Hi,

I would be interested in this too. As far as I am correct, the "normal" way of finetuning a model (or even train a new one) would be to use [this script](https://github.com/explosion/spaCy/blob/master/examples/training/train_ner.py).

Currently, If I take the sample mentioned there like:

```
# training data
TRAIN_DATA = [
    ("Wer ist Angela Merkel", {"entities": [(8, 21, "PERSON")]}),
    ("Ich mag Thomas Müller", {"entities": [(8, 21, "PERSON")]}),
]
```
I am quickly running into:

`KeyError: "[E001] No component 'pytt_tok2vec' found in pipeline. Available names: ['ner']"
`

which makes sense since actually the BERT provides: 

`['sentencizer', 'pytt_wordpiecer', 'pytt_tok2vec', 'pytt_textcat']`

but the NER pipeline expects 'tok2vec' - Am I correct @ines @honnibal ?

Is there any workaround or suggestion on how to achieve this?

**Comment by militu** [comment]
Hi, is there currently a workaround? Thank you!

**Comment by edivaldodelgado** [comment]
I was able to solve this modifing the original [script](https://github.com/explosion/spaCy/blob/master/examples/training/train_ner.py). Just commited the changes on #49

**Comment by honnibal** [comment]
@edivaldodelgado Your script won't work. The model needs to be modified to use the token vectors from the PyTorch Transformers model --- so you need to actually define a different model, and also probably subclass the parser class, so that it backpropagates into the PyTorch Transformers model.

spaCy's NER and parser components are designed to expect very narrow, not-so-detailed token vectors. We iterate over the sentence and on each step, extract a few token features, concatenate their vectors, and pass them into a feedforward model. Probably none of this makes sense when you have the very large, very detailed vectors from the BERT model.

So, it probably makes sense to just write a new component that works differently from spaCy's much more complicated NER model. Alternatively, maybe we can repurpose some of the current model, and just change some of the settings. It's not clear.

In summary: please chill :tongue:. There's a couple of weeks of work in this, including a bunch of experiments to figure out what works best. You can't just copy over a script --- the script won't actually do anything.

**Comment by edivaldodelgado** [comment]
Hi @honnibal thank you for your detailed reply =) such a great community! I actually did some changes on the script, not just copied, tried on Colab and worked (??).

The colab:
[Colab](https://colab.research.google.com/drive/1ARqGiJMkEY8ndNdcJUPAyYxzjo0GQvTd)

But now I can see that there is a fundamental difference on the structure of tokens, really don't know what happened.

**Comment by honnibal** [comment]
@edivaldodelgado You would have been training the normal spaCy CNN model, rather than using any PyTorch transformers features. This of course works --- it just doesn't do what you had in mind.

**Comment by einarbmag** [comment]
@honnibal, really exciting that you're working on a proper fine-tuning NER model, looking forward to using it! However, I thought about the simpler option of a feature-based approach as explained in the original BERT paper (resulting in a seemingly small loss of accuracy compared to finetuning), i.e. a single token-level classifier on top of the BERT representation. Would a model of this kind be considered for a PR? The speedup benefit is obviously greatest if we can pre-compute the BERT representations once, and not every time the training example is passed to the model. Any thoughts on how this could be done in a spaCy-esque way? Create a spaCy pipeline component that caches the vector representations in a hash table on disk, parse all the training data once before starting to update, and then disable the 'pytt_tok2vec' component before training?

**Comment by mmaybeno** [comment]
@honnibal I noticed a recent branch related to this surface. Is there anything we can do to help push this along to get it merged in? I'd be willing to contribute if needed.

**Comment by adrianeboyd** [comment]
Please try out `spacy-nightly` as a preview of how this is implemented in `spacy` v3.0 and `spacy-transformers` v1.0: https://nightly.spacy.io
